### ISSUE TITLE
Android pause/resume behavior

### ISSUE THREAD

**Issue by Robyt3** [ddnet/ddnet#8503]
> The entity bug.
> When I am in ddnet, whether in a server or not, if I put ddnet in the background of the phone(I dont know how to describe, just touch the menu button of the phone, or power button), when I reopen the ddnet client, all entity will turn to completely white.If you do this for several times, all tiles, editor, skins, everything will be bugged.
(Jupeyy: This should be fixed, android destroys the surface of the app while in background, the old version then tried to recreate all graphics at runtime, which didnt really work out in all cases. Our current version should instead simply leave all VRAM as is and just create/destroy the surface using SDL window destroy/create events)

Separated from https://github.com/ddnet/ddnet/issues/8359#issuecomment-2178112566.

According to https://wiki.libsdl.org/SDL2/README/android#pause--resume-behaviour we have to decide whether the game should run in the background (connection to server stays alive) or whether it is paused (less battery usage), and we have to handle `SDL_RENDER_DEVICE_RESET` at least for old hardware.

**Comment by Jupeyy** [comment]
we have to handle `SDL_RENDER_DEVICE_RESET` at least for old hardware.

this is simply such a hard task.
How exactly do you restore textures? do you want to have a list of file paths of all textures currently loaded?
Same for basically everything allocated.
Additionally member variables in all components need to be updated.
Realistically i don't see a chance to restore the gl context that way. In all my android experience which is the SDKs i targeted before you lowered it. I never seen a problem with keeping the GL context, neither in emulators nor real hardware.

**Comment by Jupeyy** [comment]
If the game should be paused is a different problem tho.

We could probably pause it in the menu, but ingame it doesn't really make sense, since it will lead to timeouts anyway. And i guess quickly tabbing out is a valid usecase. So if at all disconnect after some time and then pause.

**Comment by Robyt3** [comment]
> However, there's a chance (on older hardware, or on systems under heavy load), where the GL context can not be restored. In that case you have to listen for a specific message (SDL_RENDER_DEVICE_RESET) and restore your textures manually or quit the app.

I would simply go for "quit the app" since reloading the textures is almost equivalent to restarting the game anyway.

> We could probably pause it in the menu, but ingame it doesn't really make sense, since it will lead to timeouts anyway. And i guess quickly tabbing out is a valid usecase. So if at all disconnect after some time and then pause.

According to the documentation on `SDL_HINT_ANDROID_BLOCK_ON_PAUSE` it looks like we can't change the behavior after setting it once:

> The value should be set before SDL is initialized.

But we could reduce the update rate significantly while offline and the app is in the background if we don't block on pausing.

**Comment by Jupeyy** [comment]
> According to the documentation on `SDL_HINT_ANDROID_BLOCK_ON_PAUSE` it looks like we can't change the behavior after setting it once:

We can simply call SDL_poll_event instead of peek_event or whatever the functions are called.. one does block, the other does not.. I don't think we need an env var.. i could imagine internally it does the same anyway.

**Comment by Robyt3** [comment]
> We can simply call [SDL_WaitEvent](https://wiki.libsdl.org/SDL2/SDL_WaitEvent) instead of poll_events.. one does block, the other does not.. I don't think we need an env var.. i could imagine internally it does the same anyway.

I don't think that will work. Both `SDL_WaitEvent` and `SDL_PollEvent` may call `SDL_PumpEvents`, which would block if `SDL_HINT_ANDROID_BLOCK_ON_PAUSE` is set. The hint `SDL_HINT_ANDROID_BLOCK_ON_PAUSE` directly changes the implementation of `SDL_PumpEvents` on Android which is only done once when the android video device is initialized:

https://github.com/libsdl-org/SDL/blob/4800edfa0c997a79b3d7527d00474074466b5ad6/src/video/android/SDL_androidvideo.c#L109-L114

**Comment by Jupeyy** [comment]
> I don't think that will work. Both `SDL_WaitEvent` and `SDL_PollEvent` may call `SDL_PumpEvents`, which would block if `SDL_HINT_ANDROID_BLOCK_ON_PAUSE` is set. The hint `SDL_HINT_ANDROID_BLOCK_ON_PAUSE` directly changes the implementation of `SDL_PumpEvents` on Android which is only done once when the android video device is initialized:
> 
> https://github.com/libsdl-org/SDL/blob/4800edfa0c997a79b3d7527d00474074466b5ad6/src/video/android/SDL_androidvideo.c#L109-L114

Yes exactly, and that is why i would not set the env var at all, keep a local var that m_CanPause, and pass a new argument to our Input::Update function that is like `CanPauseIngame` and if both are true it calls WaitEvent else PollEvent (very simplified).
